### DISCUSSION TITLE
STM32RTC on STM32F401CC running fast

### DISCUSSION THREAD

**Discussion by RichK2015** [stm32duino/Arduino_Core_STM32#1333] [Q&A] [answered]
Hello all,
Re: STM32duino STM32RTC on Arduino "black pill".

I have been trying to make use of the built-in Real Time Clock logic on a STM32F401CC "Black Pill" generic module.  I have been using this version of the board for the past year and it has worked fine on the various projects I've been doing.

For this particular experiment, I downloaded the STM32duino example code from GitHub, compiled and have been running that code.  The code compiled, loaded and executed just fine on the board, however it looks like the clock seems to be gaining between 1 to 2 minutes per hour of constant running.  I haven't changed anything in the example aside for configuring the 'start time' values, and am using the STM32duino STM32RTC library version 1.03.  

My questions are:

1.  Is it necessary to tweak the code to ensure that the right clock source is in use?  I didn't see any mention of that however that doesn't mean it doesn't have to happen.  Wrong prescaler setting?
2.  I searched the STM32duino comments but didn't see anything that looked like my situation.  Is this a previously reported problem, and is there a work-around (library code change or compensation tweak I can put in the code).

I have been using a DS3231 outboard module as my RTC and would like to eliminate it if possible.

Rich

**Comment by fpistm** [comment]
Which clock you used?

**Reply by RichK2015** [reply]
Hi Fpistm,

Thanks for the response.  Since I posted I think I figured out my problem. and I'm a little ashamed I didn't dig deeper before asking one in the first place.

In the example code, it sets the clock source to be Low Speed Internal (LSI), which for the WeAct 401 board I'm using is a 38khz local RC signal.  Switching to the Low Speed External signal which apparently is the standard 38.563khz signal (?)  "rtc.setClockSource(STM32RTC::LSE_CLOCK);" brought the clock speed back into alignment.  

I suppose my laziness could prove to be of value to someone else. however since the information at least is documented in this area.

Best Regards,

Rich

**Comment by RichK2015** [comment, marked as answer]
Problem resolved by using   rtc.setClockSource(STM32RTC::LSE_CLOCK); as the clock source
